File: telemetry-interface/src/main/java/scala/meta/internal/telemetry/TelemetryService.java
package scala.meta.internal.telemetry;

public interface TelemetryService {

	void sendErrorReport(ErrorReport report);

	static final ServiceEndpoint<ErrorReport, Void> SendErrorReportEndpoint = new ServiceEndpoint<>("POST",
			"/v1/telemetry/sendErrorReport", ErrorReport.class, Void.class);

	void sendCrashReport(CrashReport report);

	static final ServiceEndpoint<CrashReport, Void> SendCrashReportEndpoint = new ServiceEndpoint<>("POST",
			"/v1/telemetry/sendCrashReport", CrashReport.class, Void.class);
}

File: telemetry-interfaces/src/main/scala/scala/meta/internal/telemetry/TelemetryService.scala
package scala.meta.internal.telemetry

import com.github.plokhotnyuk.jsoniter_scala.core._
import scala.util.Try

class FireAndForgetEndpoint[In: JsonValueCodec](
    val method: String,
    val uri: String,
) {
  def encodeInput(request: In): String = writeToString(request)
  def decodeInput(request: String): Try[In] = Try { readFromString(request) }
}

// This will be migrated to tapir endpoints in the next Commit
object TelemetryService {
  val sendErrorReportEndpoint = new FireAndForgetEndpoint[ErrorReport](
    "POST",
    "/v1/telemetry/sendErrorReport",
  )
}

trait TelemetryService {
  def sendErrorReport(errorReport: ErrorReport): Unit
}
